app/upload.go

```diff
@@ -0,0 +1,220 @@
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusRequestEntityTooLarge)
+	}
+
+	us.FileOffset = 0
+	now := time.Now()
+	us.CreateAt = model.GetMillisForTime(now)
+	us.Path = now.Format("20060102") + "/teams/noteam/channels/" + us.ChannelId + "/users/" + us.UserId + "/" + us.Id + "/" + us.Filename
+	if err := us.IsValid(); err != nil {
+		return nil, err
+	}
+
+	channel, err := a.GetChannel(us.ChannelId)
+	if err != nil {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.incorrect_channel_id.app_error",
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusBadRequest)
+	}
+	if channel.DeleteAt != 0 {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.cannot_upload_to_deleted_channel.app_error",
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusBadRequest)
+	}
+
+	us, storeErr := a.Srv().Store.UploadSession().Save(us)
+	if storeErr != nil {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.save.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+	}
+
+	return us, nil
+}
+
+func (a *App) GetUploadSession(uploadId string) (*model.UploadSession, *model.AppError) {
+	us, err := a.Srv().Store.UploadSession().Get(uploadId)
+	if err != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(err, &nfErr):
+			return nil, model.NewAppError("GetUpload", "app.upload.get.app_error",
+				nil, nfErr.Error(), http.StatusNotFound)
+		default:
+			return nil, model.NewAppError("GetUpload", "app.upload.get.app_error",
+				nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+	return us, nil
+}
+
+func (a *App) GetUploadSessionsForUser(userId string) ([]*model.UploadSession, *model.AppError) {
+	uss, err := a.Srv().Store.UploadSession().GetForUser(userId)
+	if err != nil {
+		return nil, model.NewAppError("GetUploadsForUser", "app.upload.get_for_user.app_error",
+			nil, err.Error(), http.StatusInternalServerError)
+	}
+	return uss, nil
+}
+
+func (a *App) UploadData(us *model.UploadSession, rd io.Reader) (*model.FileInfo, *model.AppError) {
+	// make sure it's not possible to upload more data than what is expected.
+	lr := &io.LimitedReader{
+		R: rd,
+		N: us.FileSize - us.FileOffset,
+	}
+	var err *model.AppError
+	var written int64
+	if us.FileOffset == 0 {
+		// new upload
+		written, err = a.WriteFile(lr, us.Path)
+		if written < minFirstPartSize && written != us.FileSize {
+			a.RemoveFile(us.Path)
+			return nil, model.NewAppError("UploadData", "app.upload.upload_data.first_part_too_small.app_error",
+				map[string]interface{}{"Size": minFirstPartSize}, "", http.StatusBadRequest)
+		}
+	} else if us.FileOffset < us.FileSize {
+		// resume upload
+		written, err = a.AppendFile(lr, us.Path)
+	}
+	if written > 0 {
+		us.FileOffset += written
+		if storeErr := a.Srv().Store.UploadSession().Update(us); storeErr != nil {
```

Comment:
If a user calls this multiple times in parallel by precisely timing their HTTP requests, wouldn't that allow them to upload arbitrarily large files?

1. User creates an upload session with file size 1000
2. User issues a request to submit 1 byte of data and transmits that 1 byte
  2.1. `us.FileOffset` is now 1
3. User issues another request to submit 1 byte of data, but pauses the request before actually sending any data over
  3.1. `us.FileOffset` doesn't get updated yet, remains at 1
4. User issues a request to submit 998 bytes and transmits all of the data
  4.1. `us.FileOffset` is now 1+998 = 999
5. User unpauses the earlier request and transmits the 1 byte of data
  5.1. `us.FileOffset` is updated again: 1+1 = 2

Repeat from step 3 until the offset reaches 1000, and the actual file size ends up being:

![image](https://user-images.githubusercontent.com/3514914/92730943-224aa400-f37d-11ea-9374-5d6314a0cd45.png)

Reply:
That sounds correct. Concurrent requests can indeed cause some consistency issues including the one you suggested.
One option, and probably the simplest to implement, could be to allow only one upload request per upload session at any given time.

Reply:
Yeah that sounds like the right approach.